toolchain/check/deduce.cpp

```diff
@@ -75,6 +75,60 @@ class DeductionWorklist {
   Context& context_;
   llvm::SmallVector<PendingDeduction> deductions_;
 };
+
+// State that is tracked throughout the deduction process.
+class DeductionContext {
+ public:
+  DeductionContext(Context& context, SemIR::LocId loc_id,
+                   SemIR::GenericId generic_id, bool diagnose)
+      : context_(&context),
+        loc_id_(loc_id),
+        generic_id_(generic_id),
+        diagnose_(diagnose) {
+    CARBON_CHECK(generic_id.is_valid(),
+                 "Performing deduction for non-generic entity");
+
+    // Initialize the deduced arguments to Invalid.
+    result_arg_ids_.resize(
+        context.inst_blocks()
+            .Get(context.generics().Get(generic_id_).bindings_id)
+            .size(),
+        SemIR::InstId::Invalid);
+  }
+
+  auto context() const -> Context& { return *context_; }
+  auto worklist() -> DeductionWorklist& { return worklist_; }
+
+  // Adds the arguments from the given specific as known enclosing generic
+  // arguments that will not be deduced.
+  auto AddEnclosingArgs(SemIR::SpecificId enclosing_specific_id) -> void;
+
+  // Performs all deductions in the deduction worklist. Returns whether
+  // deduction succeeded.
+  auto Deduce() -> bool;
+
+  // Returns whether every generic parameter has a corresponding deduced generic
+  // argument. If not, issues a suitable diagnostic.
+  auto CheckDeductionIsComplete() -> bool;
+
+  // Forms a specific corresponding to the deduced generic with the deduced
+  // argument list.
+  auto MakeSpecific() -> SemIR::SpecificId;
```

Comment:
I'm guessing this can't be called before `Deduce`? If so, it would probably be better to document that fact.

Reply:
Added comment.